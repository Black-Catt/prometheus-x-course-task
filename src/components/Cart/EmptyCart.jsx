import bigCartImg from '../../img/empty/big-cart.svg';
import Header from '../Header/Header';
import Footer from '../Footer/Footer';
import { useNavigate, Link } from 'react-router-dom';
import arrowImg from '../../img/specific-book/arrow.svg';

const EmptyCart = () => {
  const navigate = useNavigate();
  return (
    <div className="wrapper">
      <Header />
      <main className="empty">
        <div className="empty__container">
          <Link
            className="back-btn "
            onClick={() => navigate(-1)}
            to="book-list"
          >
            <img src={arrowImg} alt="back-arrow" />
          </Link>
          <div className="empty__content">
            <button disabled className="empty__btn">
              Purchase
            </button>
            <div className="empty__img">
              <img src={bigCartImg} alt="cart" />
            </div>
            <p className="empty__text">Cart is empty</p>
          </div>
        </div>
      </main>
      <Footer />
    </div>
  );
};
export default EmptyCart;
