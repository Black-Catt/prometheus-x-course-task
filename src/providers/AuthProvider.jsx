import { createContext, useState } from 'react';

export const AuthContext = createContext(null);

export const AuthProvider = ({ children }) => {
  const [user, setUser] = useState(null);
  const [userName, setUserName] = useState('');

  const signIn = (newUser, cb) => {
    setUser(newUser);
    function cb() {}
  };
  const signOut = (cb) => {
    setUser(null);
    function cb() {}
  };
  const getUserName = (userName) => {
    setUserName(userName);
  };

  const value = {
    user,
    signIn,
    signOut,
    getUserName,
    userName,
  };
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
};
