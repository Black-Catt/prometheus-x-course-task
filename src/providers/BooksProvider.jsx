import { createContext, useState } from 'react';
import data from '../data/books.json';

import imageNotFound from '../img/specific-book/imageNotFound.png';

export const BooksContext = createContext(null);

export const BooksProvider = ({ children }) => {
  const { books: bookData } = data;
  const [books, setBooks] = useState(bookData);
  const value = {
    books,
    imageNotFound,
  };

  return (
    <BooksContext.Provider value={value}>{children}</BooksContext.Provider>
  );
};
