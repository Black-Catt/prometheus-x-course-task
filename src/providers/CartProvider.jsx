import { createContext, useState } from 'react';

export const CartContext = createContext(null);

export const CartProvider = ({ children }) => {
  const [cartItems, setCartItems] = useState([]);
  const [bookData, setBookData] = useState([]);

  const addToCart = (book, quantity, total) => {
    book = Object.assign({}, book);
    const existingBook = cartItems.find((item) => item.title === book.title);

    let newCartItems;
    if (existingBook) {
      newCartItems = cartItems.map((item) => {
        if (item.title === book.title) {
          item.count = quantity;
          item.price = total;
        }
        return item;
      });
    } else {
      book.count = quantity;
      book.price = total;
      newCartItems = [...cartItems, book];
    }
    localStorage.setItem('bookQuantity', newCartItems.length);
    localStorage.setItem('books', JSON.stringify(newCartItems));
    setCartItems(newCartItems);
  };

  const clearCart = () => {
    setCartItems([]);
    setBookData([]);
    localStorage.clear();
  };

  const value = {
    cartItems,
    addToCart,
    clearCart,
    bookData,
  };

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
};
