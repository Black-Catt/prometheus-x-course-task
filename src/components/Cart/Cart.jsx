import Header from '../Header/Header';
import Footer from '../Footer/Footer';
import EmptyCart from './EmptyCart';

import useCart from '../../hooks/useCart';
import useBooks from '../../hooks/useBooks';
import { useNavigate, Link } from 'react-router-dom';
import arrowImg from '../../img/specific-book/arrow.svg';

const Cart = () => {
  const { imageNotFound } = useBooks();
  const { clearCart } = useCart();
  const navigate = useNavigate();

  const cartItems = JSON.parse(localStorage.getItem('books'));

  if (!cartItems) {
    return <EmptyCart />;
  }
  const price = cartItems.reduce((acc, el) => acc + +el.price, 0);

  return (
    <div className="wrapper">
      <Header />
      <main className="page">
        <div className="cart">
          <div className="cart__container">
            <button
              className="back-btn "
              onClick={() => navigate(-1)}
              to="book-list"
            >
              <img src={arrowImg} alt="back-arrow" />
            </button>
            <div className="cart__rows">
              {cartItems?.map((book, index) => (
                <div key={index} className="cart__row row-cart">
                  <div className="row-cart__body">
                    <div className="row-cart__book">
                      <div className="row-cart__img">
                        <img
                          src={book.image === '' ? imageNotFound : book.image}
                          alt="Book"
                        />
                      </div>
                      <p className="row-cart__name">{book.title}</p>
                    </div>
                  </div>
                  <div className="row-cart__items">
                    <div className="row-cart__count">
                      {isNaN(book.count) ? 1 : book.count}
                    </div>
                    <div className="row-cart__cost">{book.price}</div>
                  </div>
                </div>
              ))}
            </div>
            <div className="cart__bottom">
              <div className="cart__price">
                Total price, <span>{price.toFixed(2)}</span>$
              </div>
              <button onClick={() => clearCart()} className="cart__btn button">
                Purchase
              </button>
            </div>
          </div>
        </div>
      </main>
      <Footer />
    </div>
  );
};
export default Cart;
