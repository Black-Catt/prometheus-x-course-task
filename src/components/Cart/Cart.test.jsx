import { render } from '@testing-library/react';
import Cart from './Cart';
import { CartProvider } from '../../providers/CartProvider';
import { BooksProvider } from '../../providers/BooksProvider';
import { AuthProvider } from '../../providers/AuthProvider';
import { MemoryRouter } from 'react-router-dom';

describe('Cart component', () => {
  test('Renders the Cart component correctly', () => {
    const mockItems = [
      { title: 'Testing 123', image: 'image.jpg', price: 10 },
      { title: 'Test Again', image: 'image2.png', price: 15 },
    ];

    const { container } = render(
      <AuthProvider>
        <MemoryRouter>
          <BooksProvider>
            <CartProvider>
              <Cart cartItems={mockItems} />
            </CartProvider>
          </BooksProvider>
        </MemoryRouter>
      </AuthProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
