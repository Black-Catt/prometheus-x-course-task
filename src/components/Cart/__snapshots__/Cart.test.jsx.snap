// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Renders the Cart component correctly 1`] = `
<div>
  <div
    class="wrapper"
  >
    <header
      class="header"
      data-testid="header-component"
    >
      <div
        class="header__container"
        data-testid="header-container"
      >
        <h2
          class="header__tag"
          data-testid="header-tag"
        >
          X-course-task/Zaharchenko Anton
        </h2>
        <div
          class="header__actions"
        >
          <div
            class="header__menu mennu"
            data-testid="header-menu"
          >
            <div
              class="menu__body"
            >
              <a
                class="menu__cart"
                data-testid="menu-cart"
                href="/cart"
              >
                <img
                  alt="cart"
                  src="cart.svg"
                />
                <span
                  id="book-quantity"
                  style="display: none;"
                />
              </a>
              <a
                class="menu__button hidden"
                data-testid="menu-button"
                href="/"
              >
                Sign Out
              </a>
              <div
                class="menu__avatar hidden"
              >
                <img
                  alt="avatar"
                  data-testid="menu-avatar"
                  src="avatar.svg"
                />
              </div>
              <p
                class="menu__user hidden"
              />
            </div>
            <div
              class="spollers _spoller-init"
            >
              <div
                class="spollers__item"
              >
                <button
                  class="spollers__title "
                  type="button"
                >
                  <div
                    class="menu__avatar"
                  >
                    <img
                      alt="avatar"
                      src="avatar.svg"
                    />
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </header>
    <main
      class="empty"
    >
      <div
        class="empty__container"
      >
        <a
          class="back-btn "
          href="/book-list"
        >
          <img
            alt="back-arrow"
            src="arrow.svg"
          />
        </a>
        <div
          class="empty__content"
        >
          <button
            class="empty__btn"
            disabled=""
          >
            Purchase
          </button>
          <div
            class="empty__img"
          >
            <img
              alt="cart"
              src="big-cart.svg"
            />
          </div>
          <p
            class="empty__text"
          >
            Cart is empty
          </p>
        </div>
      </div>
    </main>
    <footer
      class="footer"
      data-testid="footer-component"
    >
      <div
        class="footer__container"
        data-testid="footer-container"
      >
        <p
          class="footer__copy"
          data-testid="footer-copy"
        >
          Made in
           
          <a
            data-testid="prometheus-link"
            href="https://prometheus.org.ua/"
            rel="noreferrer"
            target="_blank"
          >
            Prometheus
          </a>
           
          Â© 2022
        </p>
      </div>
    </footer>
  </div>
</div>
`;
